### PR TITLE
Make return buffers of communication blocks Modelica compatible

### PR DESCRIPTION
**Pull request by tbeu**
* Do not have the raw buffers of the read/write functions within the Modelica simulation environment. Introduce new functions for SerialPackager-based read/write functions in C code and use them from the Modelica functions.
 * Get rid of Modelica getPackage/setPackage functions of the SerialPackager
 * Adapt old data strategy in MDD_SerialPackagerSetData. This is necessary for the example models where read/write functions are within the same model and it is not defined which one must called first.
 * See #71

### CODE REVIEW

**Comment by tbeu** [comment]
2556912, 1211761 and 38112b2 actually do not belong here but are important fixes that must also be merged and avoid the conflicts with breaking changes of c273fe6.

**Comment by bernhard-thiele** [comment]
@tbeu I still need to dig a bit more into the details of your commits in order to understand it. I merged it anyway since it works and you mentioned that it would make things easier for you...

**Comment by tbeu** [comment]
OK. Feel free to ask.

File: Modelica_DeviceDrivers/Blocks/Communication.mo [hunk 1]
```diff
@@ -110,26 +107,24 @@ provided by the parameter <b>memoryID</b>. If the shared memory partition does n
     parameter Integer port_recv=10001
       "Listening port number of the server. Must be unique on the system"
       annotation (Dialog(group="Incoming data"));
-    Modelica_DeviceDrivers.Blocks.Interfaces.PackageOut pkgOut(pkg = SerialPackager(bufferSize))
-                                       annotation (Placement(transformation(
+    Modelica_DeviceDrivers.Blocks.Interfaces.PackageOut pkgOut(pkg = SerialPackager(if autoBufferSize then bufferSize else userBufferSize))
+      annotation (Placement(transformation(
```
Comment (bernhard-thiele) [inline Modelica_DeviceDrivers/Blocks/Communication.mo]: Why replacing `bufferSize` by the conditional expression `if autoBufferSize then bufferSize else userBufferSize` (here and similarly at other places)?

`bufferSize` is already defined by the conditional expression
```modelica
 when initial() then
       bufferSize = if autoBufferSize then alignAtByteBoundary(pkgOut.autoPkgBitSize) else userBufferSize;
end when;
```
hence, that seems to introduce redundant code.

**Comment by tbeu** [comment]
Yes, it is redundant in the case of evaluating the same if condition twice. But this is by design because userBufferSize is known at the time of the construction of the external objects whereas bufferSize is calculated from (connect) equations. It is the workaround to get the communication blocks running for internal solvers of SimulationX (as written in the commit message of 38112b2.

**Comment by tbeu** [comment]
Maybe it would also be helpful to set a better (= non-zero) start value for bufferSize.

**Comment by tbeu** [comment]
See the ticket [#1673](https://trac.modelica.org/Modelica/ticket/1673) and the heavy discussion there. Would be intersting if it runs in OMC.

**Comment by bernhard-thiele** [comment]
OK, guessed s.th. like that. Fine for me.

I just had quick look at [#1673](https://trac.modelica.org/Modelica/ticket/1673) and I think it is a very long example that makes it heavy to spot the important points. Just copying the two versions in two files and trying to compile using omc the first version compiles, version 2. doesn't.

**Comment by tbeu** [comment]
If V1 compiles. What is the output at time=1? 0.7 or 0.73908513321515?

You are right. I'll close this one and make a clearer example. The problem is that there are no other external objects in the MSL than the table objects.

**Comment by bernhard-thiele** [comment]
The value of V1 is constantly at 0.739085

**Comment by tbeu** [comment]
That's as expected!

**Comment by bernhard-thiele** [comment]
Generating code from omc for V2 gives the warning:

```
 Warning: OpenModelica requires that all external objects input arguments are possible to evaluate before initialization in order to avoid odd run-time failures, but {{table[1,1], table[1,2]}, {table[2,1], table[2,2]}} is a variable.
```

I don't think that waring is the reason while it doesn't compile the generate code later, but if it would compile it probably would give an runtime error...

**Comment by tbeu** [comment]
Thanks, good to know.

For sake of completeness: Both V1 and V2 give the wrong result (i.e. the start value 0.7) in SimulationX but work in Dymola.

**Comment by tbeu** [comment]
One more question on V1. You said that it is constant at 0.739085. Thus I guess you checked the result of V1.x. But what is V1.combiTimeTable1.y[1] at time=1?

**Comment by bernhard-thiele** [comment]
V1.combiTimeTable1.y[1] = 0 at all the time

**Comment by tbeu** [comment]
OK, that is bad indeed.

**Pull request merged**